src/NuGet.Core/NuGet.Packaging/PackageExtraction/PackageHelper.cs

```diff
@@ -37,16 +36,19 @@ public static bool IsPackageFile(string packageFileName, PackageSaveMode package
                 // This is to ignore archive entries that are not really files
                 return false;
             }
-            if (packageSaveMode.HasFlag(PackageSaveMode.Nuspec))
+
+            if (IsManifest(packageFileName))
```

Comment:
@yishaigalatzer the issue @MarkOsborneMS pointed out seems like a regression in the current code base. We were simply extracting the nuspec as is from the nupkg versus v2 generated a file name (https://github.com/NuGet/NuGet2/blob/eac0e51ceafc42cbebb0c66a64d15bf11e10516d/src/Core/Repositories/LocalPackageRepository.cs#L88).

We could retain the legacy behavior for v2 (packages.config) style restore and keep it version-less in the v3-style restore.